### PR TITLE
threads: add `shared` heap types

### PR DESCRIPTION
**Pull request by abrown**
This changes heap types everywhere to allow them to be `shared` as described in the shared-everything-threads [proposal]. The key refactoring is to push down all the abstract heap types (e.g., `func`, `any`, etc.) into an `AbstractHeapType`, leaving the `HeapType` enum with two variants: `Abstract` and `Concrete`. This allows the `Abstract` side to be marked `shared`; the `Concrete` side's `shared`-ness must be checked by looking at the pointed-to type, something which isn't yet addressed by this PR.

There are several parts of the PR that bump up against the boundary of what can be marked `shared` currently (e.g., `FuncType` and related types can't yet be marked `shared`). I've commented these locations with `TODO: handle shared` to be dealt with in future PRs that continue expanding `shared` everywhere.

[proposal]: https://github.com/WebAssembly/shared-everything-threads

### CODE REVIEW

File: crates/wasm-encoder/src/core/types.rs [hunk 1]
```diff
@@ -347,30 +347,46 @@ impl RefType {
     /// Alias for the `funcref` type in WebAssembly
     pub const FUNCREF: RefType = RefType {
         nullable: true,
-        heap_type: HeapType::Func,
+        heap_type: HeapType::Abstract {
+            shared: false,
+            ty: AbstractHeapType::Func,
+        },
```
Comment (abrown) [inline crates/wasm-encoder/src/core/types.rs:353]: If we think this is too wordy, I'm open to suggestions: I considered a helper like `HeapType::abstract(false, AbstractHeapType::func)` but decided to just directly construct this explicitly everywhere. But it's an option...

File: crates/wasm-mutate/src/mutators/peephole/dfg.rs [hunk 2]
```diff
@@ -758,12 +758,20 @@ impl<'a> DFGBuilder {
...
                 } => {
                     self.push_node(Lang::RefNull(RefType::Extern), idx);
                 }
                 Operator::RefNull {
-                    hty: wasmparser::HeapType::Func,
+                    hty:
+                        wasmparser::HeapType::Abstract {
+                            ty: wasmparser::AbstractHeapType::Func,
+                            ..
+                        },
```
Comment (alexcrichton) [inline crates/wasm-mutate/src/mutators/peephole/dfg.rs:774]: I think these'll technically need to match `shared: false` as well.

Also if you're up for it adding `HeapType::FUNC` might be a reasonable constant to add to avoid expanding it like this.

File: crates/wit-component/src/gc.rs [hunk 3]
```diff
@@ -1131,22 +1120,28 @@ impl Encoder {
...
+                    None => wasm_encoder::AbstractHeapType::None,
+                    NoExtern => wasm_encoder::AbstractHeapType::NoExtern,
+                    NoFunc => wasm_encoder::AbstractHeapType::NoFunc,
+                    Eq => wasm_encoder::AbstractHeapType::Eq,
+                    Struct => wasm_encoder::AbstractHeapType::Struct,
+                    Array => wasm_encoder::AbstractHeapType::Array,
+                    I31 => wasm_encoder::AbstractHeapType::I31,
+                    Exn => wasm_encoder::AbstractHeapType::Exn,
+                    NoExn => wasm_encoder::AbstractHeapType::NoExn,
+                };
```
Comment (alexcrichton) [inline crates/wit-component/src/gc.rs:1142]: I think this can be replaced with `let ty = ty.into()`

File: crates/wasmparser/src/validator/types.rs [hunk 4]
```diff
@@ -2743,63 +2743,74 @@ impl TypeList {
...
+            ) => {
+                a_shared == b_shared
+                    && match (a_ty, b_ty) {
+                        (Eq | I31 | Struct | Array | None, Any) => true,
+                        (I31 | Struct | Array | None, Eq) => true,
+                        (NoExtern, Extern) => true,
+                        (NoFunc, Func) => true,
+                        (None, I31 | Array | Struct) => true,
+                        (NoExn, Exn) => true,
+                        _ => false,
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wasmparser/src/validator/types.rs:2769]: for example one way to organize this could be:

```
(A, List | Of | As | Supertypes) => true,
(A, _) => false,
// ...
```

so that way if a new `B` is added someone would be forced to look at this and consider if `A` was a supertype/subtype of `B`

File: crates/wasm-smith/src/core/code_builder.rs [hunk 5]
```diff
@@ -6647,7 +6664,10 @@ fn extern_convert_any_valid(module: &Module, builder: &mut CodeBuilder) -> bool
             module,
             ValType::Ref(RefType {
                 nullable: true,
-                heap_type: HeapType::Any,
+                heap_type: HeapType::Abstract {
+                    shared: false, // TODO: handle shared
+                    ty: AbstractHeapType::Any,
+                },
             }),
         )
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wasm-smith/src/core/code_builder.rs:6672]: (ditto for most things above too)

File: crates/wasm-encoder/src/core/types.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/wasm-encoder/src/core/types.rs:353]: I think this is fine, it's just "the one place" to write this down for the `pub const` values

File: crates/wasmparser/src/readers/core/types.rs [hunk 6]
```diff
@@ -1454,53 +1553,14 @@ impl<'a> FromReader<'a> for RefType {
...
-                Ok(HeapType::I31)
-            }
-            0x69 => {
-                reader.read_u8()?;
-                Ok(HeapType::Exn)
-            }
-            0x74 => {
-                reader.read_u8()?;
-                Ok(HeapType::NoExn)
+            0x70 | 0x6F | 0x6E | 0x71 | 0x72 | 0x73 | 0x6D | 0x6B | 0x6A | 0x6C | 0x69 | 0x74 => {
```
Comment (alexcrichton) [inline crates/wasmparser/src/readers/core/types.rs:1561]: This reminds me that I've been meaning for some time now to clean up all the decoding here in this crate since all the various constants all over the place feel a bit overkill.

**Review by alexcrichton — COMMENTED** [review]
High-level question, how complete do you want to get this before landing? For example do you want to have all the `// TODO` resolved?

File: crates/wast/src/core/binary.rs [hunk 7]
```diff
@@ -307,69 +301,51 @@ impl<'a> Encode for HeapType<'a> {
...
+}
+
 impl<'a> Encode for RefType<'a> {
     fn encode(&self, e: &mut Vec<u8>) {
         match self {
-            // The 'funcref' binary abbreviation
-            RefType {
-                nullable: true,
-                heap: HeapType::Func,
-            } => e.push(0x70),
```
[the remark this answers is not in the document]
Comment (abrown) [reply crates/wast/src/core/binary.rs:317]: They shouldn't have, they're just grouped under a single pattern at new lines 329-332. I.e., if we have a ref that is nullable and a heap type, then print this byte encoding (using the `Encode for AbstractHeapType` in this case). I did this same kind of thing in several places. The one change in behavior from before is that, in this short-form case, we prepend `0x65` if the abstract heap type is shared.

File: crates/wasmparser/src/validator/operators.rs [hunk 8]
```diff
@@ -3991,15 +3991,23 @@ where
         let is_nullable = extern_ref
             .as_type()
             .map_or(false, |ty| ty.as_reference_type().unwrap().is_nullable());
-        let any_ref = RefType::new(is_nullable, HeapType::Any).unwrap();
+        let heap_type = HeapType::Abstract {
+            shared: false, // TODO: handle shared--is this always the case?
+            ty: AbstractHeapType::Any,
+        };
```
[the remark this answers is not in the document]
Comment (abrown) [reply crates/wasmparser/src/validator/operators.rs:3997]: Opened https://github.com/WebAssembly/shared-everything-threads/issues/65 to track this. Not sure how I can create a `HeapType::ANY` from `enum HeapType`? Some `const` form?

File: crates/wasmprinter/src/lib.rs [hunk 9]
```diff
@@ -1018,21 +1018,32 @@ impl Printer<'_, '_> {
...
-            HeapType::I31 => self.print_type_keyword("i31")?,
-            HeapType::Exn => self.print_type_keyword("exn")?,
-            HeapType::NoExn => self.print_type_keyword("noexn")?,
             HeapType::Concrete(i) => {
                 self.print_idx(&state.core.type_names, i.as_module_index().unwrap())?;
             }
+            HeapType::Abstract { shared, ty } => {
+                use AbstractHeapType::*;
+                if shared {
+                    self.start_group("shared ")?;
```
[the remark this answers is not in the document]
Comment (abrown) [reply crates/wasmprinter/src/lib.rs:1027]: Yeah, that's what it would appear that we are settling on for the long-form of shared abstract references. I.e., `(shared funcref)` == `(ref null (shared func))`.

[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wasmprinter/src/lib.rs:1027]: Ok sounds good 👍

File: crates/wast/src/core/binary.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wast/src/core/binary.rs:317]: Ah ok I think I got myself confused, carry on!

File: crates/wasmparser/src/validator/operators.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wasmparser/src/validator/operators.rs]: I think you should be able to do `impl HeapType { pub const ANY: HeapType = /* ... */; }` to get the `::ANY` working

File: crates/wasm-smith/src/core/code_builder.rs [hunk 10]
```diff
@@ -5384,18 +5378,23 @@ fn ref_null(
...
-        choices.push(r(HeapType::Array));
-        choices.push(r(HeapType::Struct));
-        choices.push(r(HeapType::I31));
-        choices.push(r(HeapType::None));
-        choices.push(r(HeapType::NoFunc));
-        choices.push(r(HeapType::NoExtern));
+        let a = |abstract_heap_type| HeapType::Abstract {
+            shared: false, // TODO: handle shared
+            ty: abstract_heap_type,
+        };
```
[the remark this answers is not in the document]
Comment (abrown) [reply crates/wasm-smith/src/core/code_builder.rs:5389]: Yeah, I figured this helper function (here and elsewhere) is a way to start down that path but clearly more will be needed once it comes time to actually turn everything on in wasm-smith.

File: crates/wasmparser/src/readers/core/types.rs [hunk 11]
```diff
@@ -857,27 +859,29 @@ impl ValType {
...
-// is stable.
+// RefType is a bit-packed enum that fits in a `u24` aka `[u8; 3]`. Note that
+// its content is opaque (and subject to change), but its API is stable.
 //
 // It has the following internal structure:
 //
 // ```
-// [nullable:u1 concrete==1:u1 index:u22]
-// [nullable:u1 concrete==0:u1 abstype:u4 (unused):u18]
+// [nullable:u1 concrete==1:u1 index:u21]
```
[the remark this answers is not in the document]
Comment (abrown) [reply crates/wasmparser/src/readers/core/types.rs]: I think you were looking at an old, un-rebased version. The latest code has this fixed.

File: crates/wasmparser/src/readers/core/types.rs (hunk 6 above)
Comment (abrown) [reply crates/wasmparser/src/readers/core/types.rs:1598]: Looks like that happened in #1601; resolving this.

File: crates/wasmparser/src/readers/core/types.rs [hunk 12]
```diff
@@ -1165,6 +1174,18 @@ impl RefType {
...
+    ///
+    /// Note that we cannot know if a concrete type is shared unless we read the
+    /// type at the index, which is not possible here (i.e., `None`).
+    pub const fn is_shared(&self) -> Option<bool> {
+        if self.is_concrete_type_ref() {
+            None
+        } else {
+            Some(self.as_u32() & Self::SHARED_BIT != 0)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (abrown) [reply crates/wasmparser/src/readers/core/types.rs:1187]: Sure. I have tried this a couple of times and it's not as easy as it might seem. How about I remove the various levels of this and switch the logic over to a `WasmModuleResources` function in another PR?

File: crates/wasmparser/src/readers/core/types.rs [hunk 13]
```diff
@@ -903,33 +907,31 @@ pub struct RefType([u8; 3]);
...
+
+                let name = ty.as_str(nullable);
+                match (nullable, shared) {
+                    // Print the shortened form if nullable; i.e., `*ref`.
+                    (true, true) => write!(f, "(shared {}ref)", name),
+                    (true, false) => write!(f, "{}ref", name),
+                    // Print the long form otherwise; i.e., `(ref *)`.
+                    (false, true) => write!(f, "(ref (shared {}))", name),
+                    (false, false) => write!(f, "(ref {})", name),
+                }
```
[the remark this answers is not in the document]
Comment (abrown) [reply crates/wasmparser/src/readers/core/types.rs:924]: I mean, what I'm trying to do here is factor out the common pattern, mainly for human comprehensibility. I would just use `RefType::wat` but (a) it's behind the `validate` feature and (b) because of its `&'static str` return it can't print out the concrete index like we were accustomed to here. If we dropped (a) and made `RefType::wat` return a `String` then we can just have one version of all this...

Perhaps better yet, why don't we just use `Debug` everywhere `RefType::wat` was being used (i.e., `ty_to_str`)?

File: crates/wast/src/core/types.rs [hunk 14]
```diff
@@ -57,11 +57,57 @@ impl<'a> Peek for ValType<'a> {
...
+        if l.peek::<Index>()? {
+            Ok(HeapType::Concrete(parser.parse()?))
+        } else if l.peek::<kw::shared>()? {
+            parser.parens(|p| {
+                p.parse::<kw::shared>()?;
+                Ok(HeapType::Abstract {
+                    shared: true,
+                    ty: p.parse()?,
+                })
+            })
```
[the remark this answers is not in the document]
Comment (abrown) [reply crates/wast/src/core/types.rs:84]: I haven't added `*.wast` tests for this PR because of the issue with checking shared-ness of concrete types. Until I can sort out how to properly check if a type is shared, I can't validate any of this. I did mention below that I wanted to try that in a separate PR, but I guess we should decide to either:
- disable `*.wast` parsing and accept this PR without tests
- fix the shared check and include the `*.wast` tests in this PR.

**Comment by abrown** [comment]
> High-level question, how complete do you want to get this before landing? For example do you want to have all the // TODO resolved?

No, I think I want to keep the boundary-level `TODO: handle shared` around so I can make sure to check all these places when I expand shared types even more. The real question for me about completeness is testing: as I describe above, I can't write spec tests without shared-checking, but I kind of wanted to defer that for a future PR. I'm open to fixing it all here, though, but it's all getting a bit unwieldy...

File: crates/wasmparser/src/readers/core/types.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wasmparser/src/readers/core/types.rs:924]: Keeping the `&str` is somewhat important for performance in the operand validator because that's used in a few places for error messages and not dealing with `Debug` or `String` helps codegen/perf there.

This may just be me not really being familiar enough with the gc proposal. I basically just keep forgetting all the various shorthands like `nullexnref` works but `(ref nullexn)` doesn't or something like that. Anyway just ignore me here I think.

**Review by alexcrichton — COMMENTED** [review]
My main worry with this is that there's no tests currently for this PR and personally at least once I land a `// TODO: ...` it's highly likely that I just forget about it and don't actually handle it until years later when someone hits a bug. I do think that some tests should be part of this, at the very least parsing/printing/dumping/etc or something along those lines. The tests don't necesarily need to be exhaustive, but I think there should be at least some simple things working at this point, even if they should be invalid modules in the future or something like that.

As for things like `is_shared`, if you're ok taking on the responsibility to audit the `// TODO: ...` after this lands I think that's ok.

File: crates/wasmparser/src/readers/core/types.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (abrown) [reply crates/wasmparser/src/readers/core/types.rs]: Done, see ad37d62.

**Review by alexcrichton — APPROVED** [review]
